libnmstate/state.py

```diff
@@ -167,6 +226,49 @@ def merge_interfaces(self, other_state):
             dict_update(other_state.interfaces[name], self.interfaces[name])
             self._ifaces_state[name] = other_state.interfaces[name]
 
+    def merge_route_config(self, current):
+        """
+        Merge routes from current to self.
+        If any changed route referring non-exist interface, create the
+        interface state with name only, in order to trigger the recreating of
+        the interface profile.
+        """
+        # Merge other_routes
+        iface_route_sets = defaultdict(set)
+        current_iface_route_sets = defaultdict(set)
+
+        for route in self.routes:
+            if Route.NEXT_HOP_INTERFACE in route:
+                iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
+                    _Route(route))
+            else:
+                logging.warning("Ignoring the route entry with no "
+                                "next hop interface defined: %s",
+                                route)
+
+        for route in current.routes:
+            iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
+                _Route(route))
+            current_iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
+                _Route(route))
+
+        for iface_name in list(six.viewkeys(iface_route_sets)):
+            # Remove routes if certain interface routes never changes.
+            if iface_route_sets[iface_name] == \
+               current_iface_route_sets.get(iface_name):
+                del iface_route_sets[iface_name]
+
+        # Create basic interface information for changed routes.
+        for iface_name in six.viewkeys(iface_route_sets):
```

Comment:
I think this method is doing more than just route state merging, it has a side affect of normalizing the interfaces state as well.
I guess you should split it and execute it explicitly in a separated method.

Reply:
New function `state.include_changed_interfaces()`